docs/src/design/changing_the_primal.md

```diff
@@ -0,0 +1,435 @@
+  y, cx = sincos(x)
+  return y, (; cx=cx)  # use a NamedTuple for the intermediates
+end
+
+pullback_at(::typeof(sin), x, y, ȳ, intermediates) = ȳ * intermediates.cx
+```
+and for the logistic sigmoid:
+```julia
+function augmented_primal(::typeof(σ), x)
+  ex = exp(x)
+  y = ex/(1 + ex)
+  return y, (; ex=ex)  # use a NamedTuple for the intermediates
+end
+
+pullback_at(::typeof(σ), x, y, ȳ, intermediates) = ȳ * y /(1 + intermediates.ex)
+```
+
+
+Cool!
+That lets us do what we wanted.
+We net decreased the time it takes to run the primal and gradient passes.
+We have now demonstrated the title question of why we needed to be able to modify the primal pass.
+We will go into that more later, and have some more examples of use.
+But first let's continue to see how we go from that `augmented_primal` to `pullback_at` to [`rrule`](@ref).
+
+
+One thing we notice when looking at `pullback_at` is it really is starting to have a lot of arguments.
+It had a fair few already, and now we are adding `intermediates` as well.
+Not to mention this is a fairly simple function, only 1 input, no keyword arguments.
+Furthermore, we don't even use all of them all the time.
+The original new code for pulling back `sin` no longer needs the `x`, and it never needed `y` (though `sigmoid` does).
+Having all this extra stuff means that the `pullback_at` function signature is long, and we are putting a bunch of extra stuff on the tape, using more memory.
+What if we generalized the idea of the `intermediate` named tuple, and had a struct that just held anything we might want put on the tape.
+```julia
+struct PullbackMemory{P, S}
+  primal_function::P
+  state::S
+end
+# convenience constructor:
+Memory(primal_function; state...) = PullbackMemory(primal_function, state)
+# convenience accessor so that `m.x` is same as `m.state.x`
+Base.getproperty(m::PullbackMemory, propname) = getproperty(getfield(m, :state), propname)
+```
+
+So changing our API we have:
+```julia
+y = f(x)  # primal program
+y, pb = augmented_primal(f, x)
+x̄ = pullback_at(pb, ȳ)
+```
+which is much cleaner.
+
+Let's try writing with this for `sin`:
+```julia
+function augmented_primal(::typeof(sin), x)
+  y, cx = sincos(x)
+  return y, PullbackMemory(sin; cx=cx)
+end
+
+pullback_at(pb::PullbackMemory{typeof(sin)}, ȳ) = ȳ * pb.cx
+```
+and for the logistic sigmoid:
+```julia
+function augmented_primal(::typeof(σ), x)
+  ex = exp(x)
+  y = ex/(1 + ex)
+  return y, PullbackMemory(σ; y=y, ex=ex)
+end
+
+pullback_at(pb::PullbackMemory{typeof(σ)}, ȳ) = ȳ * pb.y/(1 + pb.ex)
+```
+
+
+
+I think that looks pretty nice.
+
+One way we could make it look a bit nicer for usage is if the `PullbackMemory` was actually a callable object,
+since `pullback_at` only has the 2 arguments,
+and conceptually the `PullbackMemory` is a more fixed thing -- it is fully determined by the end of the primal pass,
+then during the reverse pass the argument `ȳ` is successively computed.
```

Comment:
Exactly 1 sentence per line. And can this be broken into multiple sentences?
```suggestion
One way we could make it look a bit nicer for usage is if the `PullbackMemory` was actually a callable object, since `pullback_at` only has the 2 arguments, and conceptually the `PullbackMemory` is a more fixed thing -- it is fully determined by the end of the primal pass, then during the reverse pass the argument `ȳ` is successively computed.
```

Reply:
```suggestion
One way we could make it look a bit nicer for usage is if the `PullbackMemory` was actually a callable object. `pullback_at` only has the 2 arguments.
Conceptually the `PullbackMemory` is a fixed thing it the contents of the tape for a particular operation.
It is fully determined by the end of the primal pass.
The during the gradient (reverse) pass the `PullbackMemory` is used to successively compute the `ȳ`  argument.
So it makes sense to have `PullbackMemory` being a callable object that acts on the sensitivity.
```